File: python_workflow/tests/test_hash_primitive.py
"""The keyed-hash primitive, which every deterministic-but-varying value rests on.

lexsync draws no random numbers. Anything that looks stochastic — a jittered
duration, and any future search needing a candidate order — is a pure function of
a SHA-256 digest of a key string. `hash_unit` turns that digest into a uniform
variate in [0, 1), and it has to give the *same bits* in R and Python or the two
engines will select different stimuli.

The golden digests below were frozen only after both engines were compared
bit-for-bit over 20,005 keys, including accented Spanish and CJK strings and the
empty key. R_workflow/tests/testthat/test-hash-primitive.R computes the same
digests independently, so the two suites fail together if either engine drifts.

The scheme's fragile parts, each of which this file pins:
  * 13 hex digits, not 14. Fourteen rounds up to exactly 1.0, and the integer
    mapping would then return hi + 1 with no test noticing.
  * Division by 2**52, a power of two, so the result is exact rather than close.
  * Only +, -, * and / downstream. IEEE-754 mandates those correctly rounded;
    exp, log and ** differ between R and Python by one unit in the last place.
"""
import hashlib
import struct

import pytest

from lexsync.io_utils import hash_int_range, hash_unit

TWO52 = 4503599627370496


def _bits(x: float) -> str:
    """The IEEE-754 bit pattern, so a one-bit difference cannot hide in rounding."""
    return struct.pack("<d", x)[::-1].hex()


def test_the_variate_lies_strictly_inside_the_unit_interval():
    values = [hash_unit(f"sweep|{i}") for i in range(5000)]
    assert all(0.0 <= u < 1.0 for u in values)
    # 13 hex digits cap the variate one ulp below 1, which is what stops
    # lo + floor(u * n) from ever returning hi + 1.
    assert max(values) <= 1.0 - 1.0 / TWO52


def test_known_keys_including_non_ascii():
    # These exact values are reproduced by the R suite from R's own digest call.
    assert _bits(hash_unit("sweep|0")) == "3fcd8a2fd4c97bd0"
    assert hash_int_range("42|0|A|1|niño_corazón", 200, 800) == 302
    assert hash_int_range("42|0|A|1|北京大学汉字", 200, 800) == 209


def test_the_integer_mapping_respects_both_bounds():
    for lo, hi in [(0, 1), (1, 3), (200, 800), (0, 0), (-5, 5)]:
        seen = {hash_int_range(f"k|{i}", lo, hi) for i in range(3000)}
        assert min(seen) >= lo and max(seen) <= hi
    # A degenerate range is not an error; it is a constant.
    assert hash_int_range("anything", 7, 7) == 7
    with pytest.raises(ValueError, match="hi >= lo"):
        hash_int_range("k", 10, 2)


def test_the_variate_stream_matches_the_r_engine():
    stream = "|".join(_bits(hash_unit(f"sweep|{i}")) for i in range(20000))
    assert (hashlib.sha256(stream.encode("utf-8")).hexdigest()
            == "8c1dabf515645b54c85e10529275bf0920ba965708f0d115a9c5a3b810d800f3")


def test_the_integer_stream_matches_the_r_engine():
    stream = "|".join(str(hash_int_range(f"sweep|{i}", 200, 800)) for i in range(20000))
    assert (hashlib.sha256(stream.encode("utf-8")).hexdigest()
            == "170c356c58314548ca0d2bffc31ac4da56f960d5037dc05ca78234dcb045ad97")


def test_distinct_keys_do_not_collide_in_practice():
    """Not a uniformity proof, only a guard against a broken derivation.

    A truncated or constant digest would show up immediately as a collapsed
    range, which is the failure worth catching cheaply.
    """
    values = {hash_unit(f"sweep|{i}") for i in range(20000)}
    assert len(values) == 20000

File: R_workflow/tests/testthat/test-hash-primitive.R
# The keyed-hash primitive, which every deterministic-but-varying value rests on.
#
# lexsync draws no random numbers. Anything that looks stochastic -- a jittered
# duration, and any future search needing a candidate order -- is a pure function
# of a SHA-256 digest of a key string. `hash_unit` turns that digest into a
# uniform variate in [0, 1), and it has to give the *same bits* in R and Python or
# the two engines will select different stimuli.
#
# The golden digests below were frozen only after both engines were compared
# bit-for-bit over 20,005 keys, including accented Spanish and CJK strings and the
# empty key. python_workflow/tests/test_hash_primitive.py computes the same
# digests independently, so the two suites fail together if either engine drifts.

two52 <- 4503599627370496

# The IEEE-754 bit pattern, so a one-bit difference cannot hide in rounding.
bits_of <- function(x) {
  vapply(x, function(v) paste(rev(sprintf("%02x", as.integer(writeBin(v, raw(), size = 8)))),
                              collapse = ""), character(1), USE.NAMES = FALSE)
}

test_that("the variate lies strictly inside the unit interval", {
  u <- lexsync:::hash_unit(sprintf("sweep|%d", 0:4999))
  expect_true(all(u >= 0 & u < 1))
  # 13 hex digits cap the variate one ulp below 1, which is what stops
  # lo + floor(u * n) from ever returning hi + 1.
  expect_lte(max(u), 1 - 1 / two52)
})

test_that("known keys, including non-ASCII, match the Python engine", {
  expect_equal(bits_of(lexsync:::hash_unit("sweep|0")), "3fcd8a2fd4c97bd0")
  expect_equal(lexsync:::hash_int_range("42|0|A|1|niño_corazón", 200L, 800L), 302L)
  expect_equal(lexsync:::hash_int_range("42|0|A|1|北京大学汉字", 200L, 800L), 209L)
})

# digest() hashes the stored bytes, and what R stores for an accented letter
# depends on the encoding mark and the session locale. The constants are the
# Python engine's, mirrored in test_hash_primitive.py, and the second one goes
# through .key_part and paste() the way every shuffle, balance and jitter key does.
test_that("the digest is the same bits for any encoding mark in any locale", {
  utf <- enc2utf8("caf\u00e9")
  lat <- iconv(utf, "UTF-8", "latin1")
  unk <- utf
  Encoding(unk) <- "unknown"
  expect_identical(c(Encoding(utf), Encoding(lat), Encoding(unk)),
                   c("UTF-8", "latin1", "unknown"))
  check <- function() {
    for (w in list(utf, lat, unk)) {
      expect_equal(lexsync:::hash_unit(w), 0.519767628103242)
      key <- paste("2026", lexsync:::.key_part(w), "soa", sep = "|")
      expect_equal(lexsync:::hash_unit(key), 0.7908144324504798)
    }
  }
  check()
  old <- Sys.getlocale("LC_CTYPE")
  skip_if(!isTRUE(suppressWarnings(Sys.setlocale("LC_CTYPE", "C")) == "C"),
          "cannot switch to the C locale on this platform")
  on.exit(Sys.setlocale("LC_CTYPE", old), add = TRUE)
  check()
})

test_that("the integer mapping respects both bounds", {
  for (rng in list(c(0L, 1L), c(1L, 3L), c(200L, 800L), c(0L, 0L), c(-5L, 5L))) {
    v <- lexsync:::hash_int_range(sprintf("k|%d", 0:2999), rng[1], rng[2])
    expect_gte(min(v), rng[1])
    expect_lte(max(v), rng[2])
  }
  # A degenerate range is not an error; it is a constant.
  expect_equal(lexsync:::hash_int_range("anything", 7L, 7L), 7L)
  expect_error(lexsync:::hash_int_range("k", 10L, 2L), "hi >= lo")
})

test_that("the variate stream matches the Python engine", {
  stream <- paste(bits_of(lexsync:::hash_unit(sprintf("sweep|%d", 0:19999))), collapse = "|")
  expect_equal(digest::digest(stream, algo = "sha256", serialize = FALSE),
               "8c1dabf515645b54c85e10529275bf0920ba965708f0d115a9c5a3b810d800f3")
})

test_that("the integer stream matches the Python engine", {
  v <- lexsync:::hash_int_range(sprintf("sweep|%d", 0:19999), 200L, 800L)
  stream <- paste(v, collapse = "|")
  expect_equal(digest::digest(stream, algo = "sha256", serialize = FALSE),
               "170c356c58314548ca0d2bffc31ac4da56f960d5037dc05ca78234dcb045ad97")
})

test_that("distinct keys do not collide in practice", {
  # Not a uniformity proof, only a guard against a broken derivation: a truncated
  # or constant digest would collapse the range, which is worth catching cheaply.
  u <- lexsync:::hash_unit(sprintf("sweep|%d", 0:19999))
  expect_equal(length(unique(u)), 20000L)
})
